847 record uniqueness

duncandewhurst opened the pull request:
Closes https://github.com/open-contracting/standard/issues/847

`records_reference.md`:

* Update the first paragraph to clarify the uniqueness rules for records.
* Update *Record structure* to clarify that: 
  * the releases list need only contain all releases **at the time of the record's publication**.
  * the versioned release need only contain a history of changes **up to the time of the record's publication**.
* Update *Compiled release* to clarify that the compiled release need only contain the latest values **at the time of the record's publication**.

`releases_and_records.md`

* Change 'While releases are never updated, records are updated each time there is a change.' to 'While releases are never updated, records **can be** updated each time there is a change.' to clarify that there is no requirement to update old records.

* Update `.description` in `record-package-schema.json` to describe only a record package, similar to `description` in `release-package-schema.json`. Previously, `.description` also paraphrased the description of a record in `definitions.Record.description`, possibly because there is no `record-schema.json` file. However, I think it's best to document the description of a record in only one place in the schema.

* Update `.description` in `definitions.Record`:
  * Align with the first sentence of `records_reference.md`.
  * Describe only the must/should/may rules for the releases list, compiled release and versioned release. 
  
  Previously, `.description` also paraphrased the descriptions from `.properties.releases`, `.properties.compiledRelease` and `.properties.versionedRelease`. However, I think it's best to document the descriptions of each property in only one place in the schema.

* Update `.description` in `definitions.Record.properties.releases` to clarify that the releases list need only contain all releases **at the time of the record's publication**

* Update `.description` in `definitions.Record.properties.compiledRelease` to clarify that the compiled release need only contain the latest values **at the time of the record's publication** and to replace 'latest version of all the contracting data' with 'latest value of each field'.

* Update `.description` in `definitions.Record.properties.versionedRelease` to clarify that the versioned release need only contain a history of changes **up to the time of the record's publication** and, since a versioned release can be published independently of a compiled release, replace 'all the data in the compiled release' with 'each field'.

duncandewhurst committed:
clarify uniqueness rules for records

duncandewhurst commented:
@jpmckinney, before I request a review from a helpdesk analyst, should we also document the uniqueness rules in the schema by adding the following to the description of `Record`?

> There should be a single record per contracting process per [distribution](https://www.w3.org/TR/vocab-dcat-2/#Class:Distribution), where a distribution might be a specific API endpoint or a specific bulk download file.

jpmckinney commented:
@duncandewhurst I have not reviewed, but yes, makes sense to also update the schema.

jpmckinney commented:
You'll need to merge 1.2-dev to get tests to pass.

duncandewhurst committed:
Merge branch '1.2-dev' into 847-record-uniqueness

duncandewhurst committed:
add record uniqueness rule to schema

mrshll1001 reviewed (approved):
All these changes reflect the clarifications indicated, and are very well written.

jpmckinney committed:
Merge branch '1.2-dev' into 847-record-uniqueness

jpmckinney commented on docs/schema/records_reference.md:
```diff
@@ -37,11 +37,11 @@ The following example demonstrates all package metadata and record fields.
 ## Record structure
 
-A record **must** contain an [ocid](identifiers.md#contracting-process-identifier-ocid) and all [releases](#releases) about the contracting process. As such, a record functions as an index of all releases about a contracting process.
+A record **must** contain an [ocid](identifiers.md#contracting-process-identifier-ocid) and all [releases](#releases) about the contracting process. As such, a record functions as an index of all releases about a contracting process at the time of the record's publication.
```
Move the relevant addition into the normative statement.

```suggestion
A record **must** contain an [ocid](identifiers.md#contracting-process-identifier-ocid) and all [releases](#releases) about the contracting process at the time of the record's publication. As such, a record functions as an index of releases about a contracting process.
```

jpmckinney commented on docs/schema/records_reference.md:
```diff
@@ -91,15 +91,15 @@ The considerations are:
 ### Compiled release
 
-The compiled release is the latest version of all the data about a contracting process. In other words, it provides a snapshot of the current state of a contracting process.
+The compiled release is the latest version of all the data about a contracting process at the time of the record's publication. In other words, it provides a snapshot of the current state of a contracting process at the time of the record's publication.
```
Avoid repetition:

```suggestion
The compiled release is the latest version of all the data about a contracting process at the time of the record's publication. In other words, it provides a snapshot of the contracting process at a given point in time.
```

jpmckinney commented on schema/record-package-schema.json:
```diff
@@ -156,12 +156,12 @@
         "versionedRelease": {
           "title": "Versioned release",
-          "description": "This contains the history of all the data in the compiled release.",
+          "description": "A history of changes to each field up to the time of the record's publication.",
```
Align with the wording of `compiledRelease`.

```suggestion
          "description": "The history of values of each field up to the time of the record's publication.",
```

jpmckinney commented on schema/record-package-schema.json:
```diff
@@ -156,12 +156,12 @@
         "compiledRelease": {
           "title": "Compiled release",
-          "description": "This is the latest version of all the contracting data, it has the same schema as an open contracting release.",
+          "description": "The latest value of each field at the time of the record's publication, it has the same schema as an open contracting release.",
```
Fixed pre-existing run-on sentence. The sentence isn't needed, but I leave that to https://github.com/open-contracting/standard/issues/330

```suggestion
          "description": "The latest value of each field at the time of the record's publication. It follows the same schema as a release.",
```

jpmckinney commented on docs/schema/records_reference.md:
```diff
@@ -1,6 +1,6 @@
 # Record Reference
 
-Whereas there can be multiple releases about a contracting process, there should be a single **record** per contracting process, aggregating all the releases available for the contracting process.
+A record aggregates the releases related to a contracting process. There should be a single record per contracting process per [distribution](https://www.w3.org/TR/vocab-dcat-2/#Class:Distribution), where a distribution might be a specific API endpoint or a specific bulk download file.
```
I prefer using "releases **about** a contracting process" throughout instead of **related to**, because a release is a data artifact that describes a real-world contracting process (an "about" relationship), whereas "relates to" is much more generic (e.g. a release about a contract renewal could, in a sense, "relate" to the original contract).

duncandewhurst replied:
Makes sense. Fixed in 5b055f71e421eec4796da49c07a827fcc7517162

jpmckinney commented on schema/record-package-schema.json:
```diff
@@ -122,7 +122,7 @@
     "Record": {
       "title": "Record",
-      "description": "An OCDS record must provide a list of all the existing OCDS releases related to a single contracting process and should provide a compiled release containing the current state of all fields in the release schema. An OCDS record may also provide a versioned history of all changes to the data in the compiled release.",
+      "description": "A record aggregates the releases related to a contracting process. A record must include a `releases` list, should include a `compiledRelease` and may include a `versionedRelease`. There should be a single record per contracting process per [distribution](https://www.w3.org/TR/vocab-dcat-2/#Class:Distribution), where a distribution might be a specific API endpoint or a speci ...
```
The "must" statement is enforced by `required`, which I don't think we typically repeat in field descriptions. I wonder whether the should statement can be moved to the `compiledRelease` description. The remaining "may" statement is not necessary.

duncandewhurst replied:
I guess the original intention was to have all the statements in one place. However, that need is met by `records_reference.md` so I've made the changes you suggest in 5521210ad1d01fefdd482e40064cce79cd7d4784

jpmckinney commented on schema/record-package-schema.json:
```diff
@@ -132,7 +132,7 @@
         "releases": {
           "title": "Releases",
-          "description": "All the existing OCDS releases related to the contracting process. The releases MUST be sorted into date order in the array, from oldest (at position 0) to newest (last).",
+          "description": "A list of all the OCDS releases related to the contracting process at the time of the record's publication. The releases MUST be sorted into date order in the array, from oldest (at position 0) to newest (last).",
```
We try to avoid information about data representation (data types like list or array) in the description.

I also simplified the second clause, as "position 0" might be confusing, and I also do not know how a "from ... to ..." construction could be interpreted in any other way than "from (first) to (last)".

Stylistically, we also do not capitalize normative keywords elsewhere.

```suggestion
          "description": "All the releases about the contracting process at the time of the record's publication. The releases must be sorted in date order, from oldest to newest.",
```

duncandewhurst committed:
Apply suggestions from code review

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

duncandewhurst committed:
Replace 'related to' with 'about'

duncandewhurst committed:
Update `Record` and `compiledRelease` descriptions.

Remove repeated and unnecessary statements from `Record` description. 
Move should statement to `compiledRelease` description.

jpmckinney commented:
Thanks!
